base/linalg/rowvector.jl

```diff
@@ -0,0 +1,195 @@
+"""
+    RowVector(vector)
+
+A lazy-view wrapper of an `AbstractVector`, which turns a length-`n` vector into
+a `1 x n` shaped row vector and represents the transpose of a vector (the elements
+are also transposed recursively). This type is usually constructed (and
+unwrapped) via the `transpose()` function or `.'` operator (or related
+`ctranspose()` or `'` operator).
+
+By convention, a vector can be multiplied by a matrix on it's left (`A * v`)
+whereas a row vector can be multiplied by a matrix on it's right (such that
+`v.' * A = (A.' * v).'`). It differs from a `1n`-sized matrix by the facts that
+its transpose returns a vector and the inner product `v1.' * v2` returns a
+scalar, but will otherwise behave similarly.
+"""
+immutable RowVector{T,V<:AbstractVector} <: AbstractMatrix{T}
+    vec::V
+    function RowVector(v::V)
+        check_types(T,v)
+        new(v)
+    end
+end
+
+
+@inline check_types{T1,T2}(::Type{T1},::AbstractVector{T2}) = check_types(T1, T2)
+@pure check_types{T1,T2}(::Type{T1},::Type{T2}) = T1 === transpose_type(T2) ? nothing : error("Element type mismatch. Tried to create a `RowVector{$T1}` from an `AbstractVector{$T2}`")
+
+# The element type may be transformed as transpose is recursive
+@inline transpose_type{T}(::Type{T}) = promote_op(transpose, T)
+
+# Constructors that take a vector
+@inline RowVector{T}(vec::AbstractVector{T}) = RowVector{transpose_type(T),typeof(vec)}(vec)
+@inline (::Type{RowVector{T}}){T}(vec::AbstractVector{T}) = RowVector{T,typeof(vec)}(vec)
+
+# Constructors that take a size and default to Array
+@inline (::Type{RowVector{T}}){T}(n::Int) = RowVector{T}(Vector{transpose_type(T)}(n))
+@inline (::Type{RowVector{T}}){T}(n1::Int, n2::Int) = n1 == 1 ? RowVector{T}(Vector{transpose_type(T)}(n2)) : error("RowVector expects 1×N size, got ($n1,$n2)")
+@inline (::Type{RowVector{T}}){T}(n::Tuple{Int}) = RowVector{T}(Vector{transpose_type(T)}(n[1]))
+@inline (::Type{RowVector{T}}){T}(n::Tuple{Int,Int}) = n[1] == 1 ? RowVector{T}(Vector{transpose_type(T)}(n[2])) : error("RowVector expects 1×N size, got $n")
+
+# Conversion of underlying storage
+convert{T,V<:AbstractVector}(::Type{RowVector{T,V}}, rowvec::RowVector) = RowVector{T,V}(convert(V,rowvec.vec))
+
+# similar()
+@inline similar(rowvec::RowVector) = RowVector(similar(rowvec.vec))
+@inline similar{T}(rowvec::RowVector, ::Type{T}) = RowVector(similar(rowvec.vec, transpose_type(T)))
+# There is no resizing similar() because it would be ambiguous if the result were a Matrix or a RowVector
+
+# Basic methods
+@inline transpose(vec::AbstractVector) = RowVector(vec)
```

Comment:
Since we're punting on conjugated and transposed views, this and the following transpose functions should probably make copies with `copymutable`.  That way it'll be consistent — both across conjugation and dimensionality.  `RowVector` itself can still be a lazy wrapper… it's just that transpose shouldn't be lazy until we get the other work done.